[build] add X86_64 Android support

AlwinEsch opened the pull request:
## Description

Found at https://github.com/xbmc/xbmc/pull/17363 that Android x86_64 is not yet supported.

These changes add to its construction route.

This matches with:
- https://developer.android.com/ndk/guides/standalone_toolchain
- https://developer.android.com/ndk/guides/abis.html
- https://gitlab.kitware.com/cmake/cmake/blob/v3.13.0/Modules/Platform/Android-Determine.cmake#L203

## Motivation and Context
The background is to be given when building on the other request CMake its target CPU, only this one was missing from here, therefore not correct.
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
This was tested with x86 (for 32 bit construction) and x86_64 (for 64 bit construction) from https://www.android-x86.org/
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

**EDIT:** This color bug was only on qemu.
![Bildschirmfoto vom 2020-02-15 21-26-16](https://user-images.githubusercontent.com/6879739/74594728-e2674c80-5039-11ea-847c-78a96cb88dcf.png)
Also is there a crazy bug and Will Smith looks like "The Smurfs", not sure if on Kodi or on this Android x86, but think on OS also his Icons below look not correct :thinking: (a RGB mix with R<->B)
![Schlaubi](https://user-images.githubusercontent.com/6879739/74594833-ffe8e600-503a-11ea-97d9-d524bb52de7f.jpg)
![Bildschirmfoto vom 2020-02-15 21-30-10](https://user-images.githubusercontent.com/6879739/74594792-67526600-503a-11ea-92ef-6f92852d06d5.png)





## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

AlwinEsch commented on tools/buildsteps/defaultenv:
```diff
@@ -41,8 +41,8 @@ case $XBMC_PLATFORM_DIR in
   android)
-    DEFAULT_NDK_VERSION="20"
-    DEFAULT_NDK_API="21"
+    DEFAULT_NDK_VERSION="20" # NDK package version (newer API can be inside)
```
Would it make sense to increase this to 21, Android only offers the "android-ndk-r21" and and 20 no longer supports?

Do not know if the 21 would already be available on our Jenkins.

VelocityRa replied:
Try it?

VelocityRa commented on CMakeLists.txt:
```diff
@@ -318,6 +318,11 @@ if(NOT CORE_SYSTEM_NAME STREQUAL android)
 else()
   set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} -Wl,--no-undefined")
+  if(CPU MATCHES x86_64)
+    # To become ffmpeg with assemble code.
```
```suggestion
    # Statically resolve global references to shared library (ie. ffmpeg) definitions.
```

AlwinEsch replied:
Thanks a lot! Is changed to them.

Rechi commented on docs/README.Android.md:
````diff
@@ -85,7 +85,7 @@ unzip $HOME/Downloads/sdk-tools-linux-4333796.zip -d $HOME/android-tools/android
 Extract Android NDK:
 ```
-unzip $HOME/Downloads/android-ndk-r20-linux-x86_64.zip -d $HOME/android-tools
+unzip $HOME/Downloads/android-ndk-r21-linux-x86_64.zip -d $HOME/android-tools
````
Why?

Rechi commented on docs/README.Android.md:
```diff
@@ -112,7 +112,7 @@ keytool -genkey -keystore ~/.android/debug.keystore -v -alias androiddebugkey -d
   * `open ~/android-dev.dmg`
 * Once you have your hdd image with case sensitive hfs+ file system execute all the steps inside of this filesystem. You need to adapt all paths in this guide so that they match your local environment. As an example here is a configure line that demonstrates possible paths:
-  * `./configure --with-tarballs=/Users/Shared/xbmc-depends/tarballs --host=arm-linux-androideabi --with-sdk-path=/Volumes/android-dev/android/android-sdk-macosx --with-ndk-path=/Volumes/android-dev/android/android-ndk-r20 --prefix=/Volumes/android-dev/android/xbmc-depends`
+  * `./configure --with-tarballs=/Users/Shared/xbmc-depends/tarballs --host=arm-linux-androideabi --with-sdk-path=/Volumes/android-dev/android/android-sdk-macosx --with-ndk-path=/Volumes/android-dev/android/android-ndk-r21 --prefix=/Volumes/android-dev/android/xbmc-depends`
```
Why?

Rechi commented on docs/README.Android.md:
````diff
@@ -138,17 +138,22 @@ cd $HOME/kodi/tools/depends
 Configure build for aarch64:
 ```
-./configure --with-tarballs=$HOME/android-tools/xbmc-tarballs --host=aarch64-linux-android --with-sdk-path=$HOME/android-tools/android-sdk-linux --with-ndk-path=$HOME/android-tools/android-ndk-r20 --prefix=$HOME/android-tools/xbmc-depends
+./configure --with-tarballs=$HOME/android-tools/xbmc-tarballs --host=aarch64-linux-android --with-sdk-path=$HOME/android-tools/android-sdk-linux --with-ndk-path=$HOME/android-tools/android-ndk-r21 --prefix=$HOME/android-tools/xbmc-depends
````
Why?

Rechi commented on docs/README.Android.md:
````diff
@@ -138,17 +138,22 @@ cd $HOME/kodi/tools/depends
 Or configure build for arm:
 ```
-./configure --with-tarballs=$HOME/android-tools/xbmc-tarballs --host=arm-linux-androideabi --with-sdk-path=$HOME/android-tools/android-sdk-linux --with-ndk-path=$HOME/android-tools/android-ndk-r20 --prefix=$HOME/android-tools/xbmc-depends
+./configure --with-tarballs=$HOME/android-tools/xbmc-tarballs --host=arm-linux-androideabi --with-sdk-path=$HOME/android-tools/android-sdk-linux --with-ndk-path=$HOME/android-tools/android-ndk-r21 --prefix=$HOME/android-tools/xbmc-depends
````
Why?

Rechi commented on docs/README.Android.md:
````diff
@@ -138,17 +138,22 @@ cd $HOME/kodi/tools/depends
 Or configure build for x86:
 ```
-./configure --with-tarballs=$HOME/android-tools/xbmc-tarballs --host=i686-linux-android --with-sdk-path=$HOME/android-tools/android-sdk-linux --with-ndk-path=$HOME/android-tools/android-ndk-r20 --prefix=$HOME/android-tools/xbmc-depends
+./configure --with-tarballs=$HOME/android-tools/xbmc-tarballs --host=i686-linux-android --with-sdk-path=$HOME/android-tools/android-sdk-linux --with-ndk-path=$HOME/android-tools/android-ndk-r21 --prefix=$HOME/android-tools/xbmc-depends
````
Why?

Rechi commented on tools/depends/.gitignore:
```diff
@@ -14,6 +14,7 @@
 /target/*/native/*
 /target/*/x86/*
 /target/*/x86_64-linux-gnu-*/*
+/target/*/x86*-linux-android-*/*
```
Why not x86_64-linux-android?

AlwinEsch replied:
The x86 itself was also not prevented, with them both stays away

Rechi replied:
Which x86?

Rechi commented on tools/depends/README.md:
```diff
@@ -36,13 +36,16 @@ Paths below are examples. If you want to build Kodi, follow our **[build guides]
 ### Android
 **arm**  
-`./configure --with-tarballs=$HOME/android-tools/xbmc-tarballs --host=arm-linux-androideabi --with-sdk-path=$HOME/android-tools/android-sdk-linux --with-ndk-path=$HOME/android-tools/android-ndk-r20 --prefix=$HOME/android-tools/xbmc-depends`
+`./configure --with-tarballs=$HOME/android-tools/xbmc-tarballs --host=arm-linux-androideabi --with-sdk-path=$HOME/android-tools/android-sdk-linux --with-ndk-path=$HOME/android-tools/android-ndk-r21 --prefix=$HOME/android-tools/xbmc-depends`
```
Why?

Rechi commented on tools/depends/README.md:
```diff
@@ -36,13 +36,16 @@ Paths below are examples. If you want to build Kodi, follow our **[build guides]
 
 **aarch64**  
-`./configure --with-tarballs=$HOME/android-tools/xbmc-tarballs --host=aarch64-linux-android --with-sdk-path=$HOME/android-tools/android-sdk-linux --with-ndk-path=$HOME/android-tools/android-ndk-r20 --prefix=$HOME/android-tools/xbmc-depends`
+`./configure --with-tarballs=$HOME/android-tools/xbmc-tarballs --host=aarch64-linux-android --with-sdk-path=$HOME/android-tools/android-sdk-linux --with-ndk-path=$HOME/android-tools/android-ndk-r21 --prefix=$HOME/android-tools/xbmc-depends`
```
Why?

Rechi commented on tools/depends/README.md:
```diff
@@ -36,13 +36,16 @@ Paths below are examples. If you want to build Kodi, follow our **[build guides]
 
 **x86**  
-`./configure --with-tarballs=$HOME/android-tools/xbmc-tarballs --host=i686-linux-android --with-sdk-path=$HOME/android-tools/android-sdk-linux --with-ndk-path=$HOME/android-tools/android-ndk-r20 --prefix=$HOME/android-tools/xbmc-depends`
+`./configure --with-tarballs=$HOME/android-tools/xbmc-tarballs --host=i686-linux-android --with-sdk-path=$HOME/android-tools/android-sdk-linux --with-ndk-path=$HOME/android-tools/android-ndk-r21 --prefix=$HOME/android-tools/xbmc-depends`
```
Why?

Rechi commented on tools/depends/configure.ac:
```diff
@@ -264,7 +264,12 @@ case $host in
       i*86*-linux-android*)
         if test "x$use_cpu" = "xauto"; then
-          use_cpu=$host_cpu
+          use_cpu=i686
```
Why?

Rechi replied:
I still do not understand why this change is necessary.

AlwinEsch replied:
Was there on https://developer.android.com/ndk/guides/standalone_toolchain:

___Table 2. Specifiable -march values and their resulting targets.___

Toolchain | Location
-- | --
arm | $TOOLCHAIN/**arm**-linux-androideabi/lib/
arm64 | $TOOLCHAIN/**aarch64**-linux-android/lib/
x86 | $TOOLCHAIN/**i686**-linux-android/lib/
x86_64 | $TOOLCHAIN/**x86_64**-linux-android/lib/

But you seems right, there also this:
```
Selecting Your Toolchain
Before anything else, you need to decide which processor architecture your standalone toolchain is going to target. This is done with the --arch flag.
```

Rechi replied:
What do you think is the value of `$host_cpu`, if you configure with `--host=i686-linux-android`?

AlwinEsch replied:
Becomes changed seen also others are possible, see note above "But you seems right, there also this:"

Rechi reviewed (changes requested):
Which devices are running Android on an x86_64 CPU? How many exists? Who will run Kodi on those?

Who will maintain x86_64-linux-android?

Why are all those unrelated changes in this PR?

AlwinEsch commented:
How much sense makes it to support 32bit x86 and not 64Bit? How much they use no idea, but to support half makes no sense.
Do you want every commit in a own request, they are all related to build and android.
And why r21, that’s a very easy answer r20 since long time no more supported from Android.
Also can i*86*-linux-android always only i686-Linux-android nothing else and only 32bit.

AlwinEsch commented:
And please come not only with „why“ I fell me like a stupid man who become a hammer on the head! Or not know what 1+1 is.

Rechi commented:
There are no official `i686-linux-android` release of Kodi since 17.3, because there isn't a relevant number of devices. The same is true for `x86_64-linux-android`. If you add the support for it people expect it to work.
I don't want commits an which change things unrelated to the commit message or only include half of the changes required.
Now `docs/README.Android.md` is inconsistent and also CI still uses NDK 20. So should everyone now compile with NDK 21 locally to know it is still working?
`Why?` just means why it this change needed, because I don't get the reason for those?

AlwinEsch commented:
Then it should be removed complete and not to leave dead to confuse everyone. If it stay it then should be this also there, to have half is bad.

 Can you give me a link to the CI that I see it. Or you mean by CI our Jenkins?

Rechi commented:
https://jenkins.kodi.tv/view/Android/
Not every commit needs a separate PR, but things that are likely to cause issues, shouldn't be hidden in unrelated PRs.

AlwinEsch commented:
Is updated and the version change removed, also the note is in separate request.
Find it with x86_64 in it more correct than before, only halfway with i686 alone.

Throwing it out would probably be the wrong way, even if only rarely used, then it would be better to point out somewhere that it is not looked after and maintained.

Maybe we could do that in the readme's?:

> Note: Android i686 and x86_64 are not maintained and are not 100% sure that everything works correctly!

AlwinEsch commented:
@Rechi would it make sense and is OK to add this "Note: ..." to Readme's?

AlwinEsch committed:
[docs] add note to README's that Android x86 and x86_64 not maintained

AlwinEsch committed:
[addons] add platform for android-x86_64 binary addons

Without are for them created binary addons not recognized.

AlwinEsch commented:
Is updated and now also tested with some binary addons.

From my side finished.

Rechi commented on CMakeLists.txt:
```diff
@@ -318,6 +318,11 @@ if(NOT CORE_SYSTEM_NAME STREQUAL android)
+  if(CPU MATCHES x86_64)
+    # Statically resolve global references to shared library (ie. ffmpeg) definitions.
+    # Related to https://stackoverflow.com/questions/46307266/including-objects-to-a-shared-library-from-a-c-archive-a
+    set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} -Wl,-Bsymbolic")
```
I'm not sure if this is the correct place. Wouldn't that apply everywhere?

AlwinEsch replied:
I don't like it there either, unfortunately Kodi is created on Android as shared libkodi.so and must unfortunately be used on it.

This annoys me too, that ffmpeg does not fix this. There have already been various users inquiries about where ffmpeg was referred to use this flag.

AlwinEsch replied:
Found to use this `set_target_properties(${APP_NAME_LC} PROPERTIES LINK_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} -Wl,-Bsymbolic")` to make sure that really only on libkodi.so and nowhere else on his CMakeLists.txt becomes used.

Rechi replied:
I thought of adding it as default linker flags everywhere, not only for the Kodi project or the kodi binary. Or could this cause different (linking) behaviour?

AlwinEsch replied:
During test it was mandatory to have on Kodi only, if this was e.g. on ffmpeg itself, becomes it to bad interact between both (compile works normal) and Kodi is not executable-

Rechi commented on tools/buildsteps/android-x86_64/make-native-depends:
```diff
@@ -0,0 +1,9 @@
+then
+  git clean -xffd $WORKSPACE/tools/depends/native
+  cd $WORKSPACE/tools/depends/native;make -j $BUILDTHREADS && tagSuccessFulBuild $WORKSPACE/tools/depends
+fi
```
Add a new line at the end of the file.

Rechi commented on tools/depends/configure.ac:
```diff
@@ -267,6 +267,11 @@ case $host in
+        if test "x$use_cpu" = "xauto"; then
+          use_cpu=$host_cpu
+        fi
+      ;;
```
Any reason to not combine this into the above switch case? The body is exactly the same.

AlwinEsch replied:
Not really :smile:, thought on begin it would be better to have separated (in case different flags needed), but you are completely right. Make no sense to have separate. Becomes changed.

AlwinEsch committed:
[build] add android x86_64 to buildstep scripts

AlwinEsch committed:
[build][depends] add support for x86_64-linux-android

Previously, only an "i686-linux-android" was used which actually only corresponds to
the 32bit CPU.

In reference to this https://developer.android.com/ndk/guides/other_build_systems at 64bit "x86_64-linux-android" is used.

AlwinEsch commented on tools/android/packaging/Makefile.in:
```diff
@@ -27,7 +27,9 @@ endif
-ifeq ($(findstring x86,$(CPU)),x86)
+ifeq ($(findstring x86_64,$(CPU)),x86_64)
+  ARCH=x86_64
+else ifeq ($(findstring x86,$(CPU)),x86)
```
As note: This change is new added.

Found during last review the x86/x86_64 also here checked.

AlwinEsch commented:
jenkins build this with addons please

Pentaphon commented:
Does this mean Nexus Player users will have working Kodi 19 builds when 19 is released?

AlwinEsch commented:
Not directly, unfortunately there are not enough people who maintain and test the Android on x86 / x86_64, so this only allows you to build your own. But will not be included in the Android store.

Will address it to the team, but don't think anything will change.

Pentaphon commented:
I think we will see a need for this as x64 makes a comeback to Android (or Android's successor, even!) in order to compete with ARM64 so I predict your commit will prove valuable down the road.
